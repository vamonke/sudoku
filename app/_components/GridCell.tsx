import { Cell } from "@/types";
import classNames from "classnames";

type Props = {
  index: number;
  cell: Cell;
  showConflict: boolean;
  onChangeCell: (index: number, value: number | null) => void;
  onFocus: (index: number) => void;
  onBlur: () => void;
  isComplete: boolean;
  isSelected: boolean;
  isHighlighted: boolean;
  hasConflict: boolean;
};

const GridCell = (props: Props) => {
  const {
    index,
    cell: { value, editable },
    onChangeCell,
    onFocus,
    onBlur,
    showConflict,
    isComplete,
    isSelected,
    isHighlighted,
    hasConflict,
  } = props;

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (/^ArrowUp|ArrowDown|ArrowLeft|ArrowRight$/.test(e.key)) {
      let nextIndex = null;
      const x = index % 9;
      const y = Math.floor(index / 9);
      switch (e.key) {
        case "ArrowLeft":
          nextIndex = x === 0 ? index + 8 : index - 1;
          break;
        case "ArrowRight":
          nextIndex = x === 8 ? index - 8 : index + 1;
          break;
        case "ArrowUp":
          nextIndex = y === 0 ? index + 72 : index - 9;
          break;
        case "ArrowDown":
          nextIndex = y === 8 ? index - 72 : index + 9;
          break;
      }
      if (nextIndex !== null && nextIndex >= 0 && nextIndex < 81) {
        const nextCell = document.querySelector(
          `input:nth-child(${nextIndex + 1})`
        ); // Is there a better way to do this?
        if (nextCell) (nextCell as HTMLElement).focus();
      }
    }
    if (!/^[1-9]|Backspace|Tab$/.test(e.key)) {
      e.preventDefault();
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    if (value === "") {
      onChangeCell?.(index, null);
    } else {
      const parsedInt = parseInt(value[value.length - 1]);
      onChangeCell?.(index, parsedInt);
    }
  };

  const hasConflictClass = hasConflict && showConflict;

  let bgClassName = "bg-gray-200";
  if (isComplete) {
    bgClassName += "bg-green-100";
  } else {
    if (hasConflictClass) {
      bgClassName = "bg-red-100";
    } else if (isSelected) {
      bgClassName = "bg-blue-200";
    } else if (isHighlighted) {
      bgClassName = "bg-blue-100";
    }
  }

  const className = classNames(
    "flex items-center justify-center text-center",
    editable && (hasConflictClass ? "text-red-400" : "text-blue-400"),
    !editable && "select-none",
    bgClassName
  );

  return (
    <input
      className={className}
      value={value ?? ""}
      onKeyDown={handleKeyDown}
      onChange={handleChange}
      readOnly={!editable}
      onFocus={() => onFocus(index)}
      onBlur={() => onBlur()}
      type="number"
      min={1}
      max={9}
      step={1}
    />
  );
};

export default GridCell;
