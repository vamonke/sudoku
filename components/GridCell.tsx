import { Cell } from "@/types";
import classNames from "classnames";
import { useEffect, useRef } from "react";

type Props = {
  index: number;
  cell: Cell;
  onChangeCell: (index: number, value: number | null) => void;
  onFocus: () => void;
  onBlur: () => void;
  isComplete: boolean;
  isSelected: boolean;
  isHighlighted: boolean;
  hasConflict: boolean;
  onArrowKey: (index: number, key: string) => void;
};

const GridCell = (props: Props) => {
  const {
    index,
    cell: { value, editable },
    onChangeCell,
    onFocus,
    onBlur,
    isComplete,
    isSelected,
    isHighlighted,
    hasConflict,
    onArrowKey,
  } = props;

  const inputRef = useRef<HTMLInputElement>(null);

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    e.preventDefault();
    if (/^ArrowUp|ArrowDown|ArrowLeft|ArrowRight$/.test(e.key)) {
      onArrowKey(index, e.key);
    }
    if (!editable) return;
    if (/^[1-9]$/.test(e.key)) {
      const parsedInt = parseInt(e.key);
      onChangeCell(index, parsedInt);
    }
    if (/^Backspace$|Delete$/.test(e.key)) {
      onChangeCell(index, null);
    }
  };

  let bgClassName = "bg-gray-200";
  if (isComplete) {
    bgClassName += "bg-green-100";
  } else {
    if (hasConflict) {
      bgClassName = "bg-red-100";
    } else if (isSelected) {
      bgClassName = "bg-blue-200";
    } else if (isHighlighted) {
      bgClassName = "bg-blue-100";
    }
  }

  const className = classNames(
    "flex items-center justify-center text-center",
    editable && (hasConflict ? "text-red-400" : "text-blue-400"),
    !editable && "select-none",
    bgClassName
  );

  useEffect(() => {
    if (isSelected && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isSelected]);

  return (
    <input
      id={`cell-${index}`} // for debugging
      ref={inputRef}
      data-testid="cell"
      className={className}
      value={value ?? ""}
      onKeyDown={handleKeyDown}
      onChange={(e) => e.preventDefault()}
      readOnly={!editable}
      onFocus={onFocus}
      onBlur={onBlur}
      type="number"
      min={1}
      max={9}
      step={1}
    />
  );
};

export default GridCell;
