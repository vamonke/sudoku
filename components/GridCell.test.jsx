import "@testing-library/jest-dom";
import { render, fireEvent, screen } from "@testing-library/react";
import GridCell from "./GridCell";

const filledCell = { value: 5, editable: true };
const emptyCell = { value: null, editable: true };
const uneditableCell = { value: 1, editable: false };

describe("GridCell", () => {
  describe("renders correctly", () => {
    it("when cell value is null", () => {
      const { container } = render(<GridCell cell={emptyCell} />);
      expect(container).toMatchSnapshot();
    });

    it("when cell value is a number", () => {
      const { container } = render(<GridCell cell={filledCell} />);
      expect(container).toMatchSnapshot();
    });

    it("when cell is uneditable", () => {
      const { container } = render(<GridCell cell={uneditableCell} />);
      expect(container).toMatchSnapshot();
    });

    it("when cell is selected", () => {
      const { container } = render(<GridCell cell={filledCell} isSelected />);
      expect(container).toMatchSnapshot();
    });

    it("when cell is highlighted", () => {
      const { container } = render(
        <GridCell cell={filledCell} isHighlighted />
      );
      expect(container).toMatchSnapshot();
    });

    it("when editable cell has conflict", () => {
      const { container } = render(<GridCell cell={filledCell} hasConflict />);
      expect(container).toMatchSnapshot();
    });

    it("when unedtiable cell has conflict", () => {
      const { container } = render(
        <GridCell cell={uneditableCell} hasConflict />
      );
      expect(container).toMatchSnapshot();
    });

    it("when cell is complete", () => {
      const { container } = render(<GridCell cell={filledCell} isComplete />);
      expect(container).toMatchSnapshot();
    });
  });

  it("calls onBlur when blurred", () => {
    const onBlur = jest.fn();
    render(<GridCell cell={filledCell} onBlur={onBlur} />);
    fireEvent.blur(screen.getByTestId("cell"));
    expect(onBlur).toHaveBeenCalled();
  });

  it("calls onFocus when focused", () => {
    const onFocus = jest.fn();
    render(<GridCell cell={filledCell} onFocus={onFocus} />);
    fireEvent.focus(screen.getByTestId("cell"));
    expect(onFocus).toHaveBeenCalled();
  });

  it("calls onChangeCell when value is changed to a number", () => {
    const onChangeCell = jest.fn();
    render(<GridCell index={0} cell={emptyCell} onChangeCell={onChangeCell} />);
    fireEvent.change(screen.getByTestId("cell"), { target: { value: 6 } });
    expect(onChangeCell).toHaveBeenCalledWith(0, 6);
  });

  it("calls onChangeCell when value is changed to empty", () => {
    const onChangeCell = jest.fn();
    render(
      <GridCell index={0} cell={filledCell} onChangeCell={onChangeCell} />
    );
    fireEvent.change(screen.getByTestId("cell"), { target: { value: "" } });
    expect(onChangeCell).toHaveBeenCalledWith(0, null);
  });

  it("should call onArrowKey when arrow key is pressed", () => {
    const onArrowKeyMock = jest.fn();
    const { getByTestId } = render(
      <GridCell index={0} cell={emptyCell} onArrowKey={onArrowKeyMock} />
    );
    const inputElement = getByTestId("cell");
    fireEvent.keyDown(inputElement, { key: "ArrowUp" });
    expect(onArrowKeyMock).toHaveBeenCalledWith(0, "ArrowUp");

    fireEvent.keyDown(inputElement, { key: "ArrowDown" });
    expect(onArrowKeyMock).toHaveBeenCalledWith(0, "ArrowDown");

    fireEvent.keyDown(inputElement, { key: "ArrowRight" });
    expect(onArrowKeyMock).toHaveBeenCalledWith(0, "ArrowRight");

    fireEvent.keyDown(inputElement, { key: "ArrowLeft" });
    expect(onArrowKeyMock).toHaveBeenCalledWith(0, "ArrowLeft");
  });
});
