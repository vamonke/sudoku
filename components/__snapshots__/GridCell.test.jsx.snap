// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridCell renders correctly when cell is complete 1`] = `
<div>
  <input
    class="flex items-center justify-center text-center text-blue-400 bg-gray-200bg-green-100"
    data-testid="cell"
    id="cell-undefined"
    max="9"
    min="1"
    step="1"
    type="number"
    value="5"
  />
</div>
`;

exports[`GridCell renders correctly when cell is highlighted 1`] = `
<div>
  <input
    class="flex items-center justify-center text-center text-blue-400 bg-blue-100"
    data-testid="cell"
    id="cell-undefined"
    max="9"
    min="1"
    step="1"
    type="number"
    value="5"
  />
</div>
`;

exports[`GridCell renders correctly when cell is selected 1`] = `
<div>
  <input
    class="flex items-center justify-center text-center text-blue-400 bg-blue-200"
    data-testid="cell"
    id="cell-undefined"
    max="9"
    min="1"
    step="1"
    type="number"
    value="5"
  />
</div>
`;

exports[`GridCell renders correctly when cell is uneditable 1`] = `
<div>
  <input
    class="flex items-center justify-center text-center select-none bg-gray-200"
    data-testid="cell"
    id="cell-undefined"
    max="9"
    min="1"
    readonly=""
    step="1"
    type="number"
    value="1"
  />
</div>
`;

exports[`GridCell renders correctly when cell value is a number 1`] = `
<div>
  <input
    class="flex items-center justify-center text-center text-blue-400 bg-gray-200"
    data-testid="cell"
    id="cell-undefined"
    max="9"
    min="1"
    step="1"
    type="number"
    value="5"
  />
</div>
`;

exports[`GridCell renders correctly when cell value is null 1`] = `
<div>
  <input
    class="flex items-center justify-center text-center text-blue-400 bg-gray-200"
    data-testid="cell"
    id="cell-undefined"
    max="9"
    min="1"
    step="1"
    type="number"
    value=""
  />
</div>
`;

exports[`GridCell renders correctly when editable cell has conflict 1`] = `
<div>
  <input
    class="flex items-center justify-center text-center text-red-400 bg-red-100"
    data-testid="cell"
    id="cell-undefined"
    max="9"
    min="1"
    step="1"
    type="number"
    value="5"
  />
</div>
`;

exports[`GridCell renders correctly when unedtiable cell has conflict 1`] = `
<div>
  <input
    class="flex items-center justify-center text-center select-none bg-red-100"
    data-testid="cell"
    id="cell-undefined"
    max="9"
    min="1"
    readonly=""
    step="1"
    type="number"
    value="1"
  />
</div>
`;
